File: 8.Sorting/GoLang/3_selection_sort.go
package main

import "fmt"

func getMinIndex(arr []int, start int) int {
	i := start
	index := i
	for i < len(arr) {
		if arr[i] < arr[index] {
			index = i
		}
		i++
	}
	return index
}

func swap(arr []int, x int, y int) {
	temp := arr[x]
	arr[x] = arr[y]
	arr[y] = temp
}

func selSort(arr []int) {
	for i := 0; i < len(arr); i++ {
		min := getMinIndex(arr, i)
		swap(arr, i, min)
	}
}

func show(arr []int, msg string) {
	println(msg)
	fmt.Printf("%v", arr)
}

func main() {
	arr := []int{22, 1, 9, 13, 2, 0, 12}
	show(arr, "before")
	selSort(arr)
	show(arr, "after")
}

File: 8.Sorting/C++/3_selection_sort.cpp
#include <iostream>
#include <string>

using namespace std;

void show(int arr[], int size, string msg) {
    cout << msg << endl;
    for (int i=0; i<size; i++) {
        cout << arr[i] << " ";
    }
    cout << endl;
}

int getMinIndex(int arr[], int start, int end) {
    int i = start;
    int index = i + 1;
    while (i < end) {
        if (arr[i] < arr[index]) {
            index = i;
        }
        i++;
    }
    return index;
}

void swap(int arr[], int x, int y) {
    int temp = arr[x];
    arr[x] = arr[y];
    arr[y] = temp;
}

void selSort(int arr[], int size) {
    for (int i=0; i<size-1; i++) {
        int min = getMinIndex(arr, i, size);
        swap(arr, i, min);
    }
}

int main() {
    int arr[] = {128, 1, 18, 42, 0, 11, 22};
    int size = sizeof(arr) / sizeof(int);
    show(arr, size, "before");
    selSort(arr, size);
    show(arr, size, "after");
}